### PR TITLE
Add inline to every String function

### LINKED ISSUE #20822: Mark String::as_mut_vec inline
as_mut_vec is free but doesn't get inlined because it's not marked inline.

### PR DESCRIPTION
**Pull request by mahkoh**
Closes #20822

### CODE REVIEW

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -138,6 +138,7 @@ impl String {
     /// let output = String::from_utf8_lossy(input);
     /// assert_eq!(output.as_slice(), "Hello \u{FFFD}World");
     /// ```
+    #[inline]
```
Comment (pnkfelix) [inline src/libcollections/string.rs]: I'm not convinced this should be marked `inline`.

File: src/libcollections/string.rs [hunk 2]
```diff
@@ -276,6 +277,7 @@ impl String {
     /// v[4] = 0xD800;
     /// assert!(String::from_utf16(v).is_err());
     /// ```
+    #[inline]
```
Comment (pnkfelix) [inline src/libcollections/string.rs]: this is easier to swallow than `fn from_utf8_lossy`, but again, not as trivial as the other methods marked in this patch

File: src/libcollections/string.rs (hunk 1 above)
Comment (mahkoh) [inline src/libcollections/string.rs]: If there were an attribute #[can_be_inlined] I would have used that. Adding #[inline] is the only way to let llvm decide if it should be inlined.

**Comment by nikomatsakis** [comment]
I think I agree with @pnkfelix; the one-line functions seem like no brainers, those that contain non-trivial loops, less clearly so. I think I'm inclined to leave the `#[inline]` of those fns for now. (Do you have an example of some real program that is slower as a result or something?)

**Comment by mahkoh** [comment]
Updated

**Pull request merged**